lib/bolt/module_installer/puppetfile.rb

```diff
@@ -97,20 +97,43 @@ def assert_satisfies(specs)
           end
         end
 
-        return if unsatisfied_specs.empty?
-
+        #versionless_mods = @modules.any? do |mod|    
+        #  mod.version.nil? 
+        #end
+        versionless_mods = []
+        @modules.each {|mod|
+          if mod.version.nil?
+            versionless_mods << mod
+          end
+        }
+        
+        return if unsatisfied_specs.empty? && versionless_mods.empty?
         command = Bolt::Util.windows? ? 'Install-BoltModule -Force' : 'bolt module install --force'
+        
+        if !unsatisfied_specs.empty?
+          message = <<~MESSAGE.chomp
+            Puppetfile does not include modules that satisfy the following specifications:
+
+            #{unsatisfied_specs.map(&:to_hash).to_yaml.lines.drop(1).join.chomp}
+            
+            This Puppetfile might not be managed by Bolt. To forcibly overwrite the
+            Puppetfile, run '#{command}'.
+          MESSAGE
+
+          raise Bolt::Error.new(message, 'bolt/missing-module-specs')
+
+        else
```

Comment:
```suggestion
        if versionless_mods.any?
```

Having this be a separate conditional instead of an `else` clause makes the code a little clearer. It'll also be needed if the `return` statement above is deleted.